block-storage/src/main/scala/coop/rchain/blockstorage/dag/BlockDagKeyValueStorage.scala

```diff
@@ -133,57 +134,116 @@ final class BlockDagKeyValueStorage[F[_]: Concurrent: Log] private (
+        .flatMap(blockMetadataIndex.getUnsafe)
+        .map(_.seqNum)
+      val checkRegressiveSeqNum = senderCurLMSeqNum.map(block.seqNum < _)
+
+      // Genesis block has empty sender. TODO Why not have list of senders?
+      val senderIsEmpty = block.sender == ByteString.EMPTY
+
+      for {
+        // This filter is required to enable adding blocks backward from higher height to lower
+        newlyBondedUnseen <- newlyBondedSet.toList.filterA(
+                              lm => latestMessagesIndex.contains(lm).not
+                            )
+        newlyBondedLMs = newlyBondedUnseen.map(v => (v, block.blockHash)).toSet
+        // Genesis children blocks senders won't have themselves in justification list, but they are not newly bonded.
+        senderHasLM         <- latestMessagesIndex.contains(block.sender)
+        senderIsNewlyBonded = !senderHasLM && newlyBondedSet.contains(block.sender)
+        senderIsNotBonded   = !bonds(block).exists(_.validator == block.sender)
+
+        blockIsGenesisChild = senderHasLM && newlyBondedSet.contains(block.sender)
+
+        //Do not create LM for empty sender, which is a now valid sender for genesis
+        lms <- if (senderIsNewlyBonded || senderIsEmpty)
+                newlyBondedLMs.pure[F]
+              // Special case when sender is bonded in genesis, so it has latest message but is looks like newly bonded
+              // Because it does not have self-justification (genesis has empty sender)
+              // TODO again - not really consistent experience
+              else if (blockIsGenesisChild)
+                (newlyBondedLMs + sendersNewLM).pure[F]
+              // This check is a hack required for some tests to work, as they use empty bonds map for genesis or
+              // validator that is not in bonds map. In this case we have to create latest message for this validator
+              // But in general - this should be cut at the stage of block verification, so its subject to an exception
+              else if (senderIsNotBonded && !senderIsEmpty)
+                (newlyBondedLMs + sendersNewLM).pure[F]
+              else
+                // Update LatestMessage for sender
+                for {
+                  _ <- Sync[F]
+                        .raiseError[(ByteString, ByteString)](BlockSenderIsMalformed(block))
+                        // genesis has empty sender, so we have to put additional check
+                        .whenA(!senderHasValidFormat && !senderIsEmpty)
+                  _ <- logEmptySender.whenA(senderIsEmpty)
+                  // As we can add blocks in backwards (e.g. when reconstructing finalized state) -
+                  // to ot overwrite existing LMs, we should have this check
+                  seqNumIsRegressive <- checkRegressiveSeqNum
+                  r = if (seqNumIsRegressive)
+                    newlyBondedLMs
+                  else
+                    newlyBondedLMs + sendersNewLM
```

Comment:
Here is a situation where two things can happen (not counting error a few lines above), `sendersNewLM` will be added to `newlyBondedLMs` or not.
So to be clear it would be nice to have some king of "logical table" where is more visible the dependency between expressions.

E.g. `senderIsEmpty` is basically on top, if we don't have a sender all bets are off to add it to the latest messages, so it make sense to use it only in the last definition like:
`val shouldAddToLatest = senderIsEmpty && allOtherChecks`.
For example now I'm not sure it `checkRegressiveSeqNum` should be executed if sender is empty.

Reply:
Yeah, all this is quite cumbersome. But thats because of that super special genesis block, which I was hoping to refactor at some point. But I agree, let me think a bit on this again.